### PR TITLE
feat/icrc-2-integration-and ICRC-1-enhancements

### PR DESCRIPTION
**Pull request by Julio-SDSharp**
# ICRC 2 Integration and ICRC 1 Enhancements

This pull request introduces ICRC 2 functionality and enhances existing ICRC 1 features, enabling:

- Transaction management: Approve and manage ICRC 2 transactions
- Allowance management: View and manage allowances granted to you or created by you
- Expanded transfer options: Transfer funds from various sources and to different destinations

## Key Features

- Enhanced Contact Book: See contacts who granted you allowances
- Allowance Management UI: Easily view and manage created and received allowances
- Flexible Transfers: Transfer funds from subaccounts, allowances, and to own account or third parties

## Addtional notes

- [ICRC-2](https://github.com/dfinity/ICRC-1/blob/main/standards/ICRC-2/README.md)

### CODE REVIEW

File: frontend/helpers/icrc.ts [hunk 1]
```diff
@@ -43,21 +44,103 @@ export async function ICRCApprove(params: ApproveParams, assetAddress: string):
...
+      newSubAccounts.push({
+        ...subAccounts[subAccountIndex],
+        allowance: response,
+      });
+    } else {
+      newSubAccounts.push(subAccounts[subAccountIndex]);
+    }
+  }
+
+  console.log(newSubAccounts);
```
Comment (OliverSDSharp) [inline frontend/helpers/icrc.ts]: delete

File: frontend/pages/contacts/components/AddContact/SubAccountFormItem.tsx [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                </div>
+              );
+            })}
+          </div>
+        </div>
+      </div>
+    </div>
+  );
+
+  function isAvailableAddContact() {
```
Comment (OliverSDSharp) [inline frontend/pages/contacts/components/AddContact/SubAccountFormItem.tsx]: can we reduce this component? maybe put those functions on other file or hook?

File: frontend/pages/contacts/components/addContact.tsx [hunk 3]
```diff
@@ -41,47 +52,56 @@ const AddContact = ({ setAddOpen }: AddContactProps) => {
     setNewContactSubNameErr,
     newContactSubIdErr,
     setNewContactSubIdErr,
-    checkPrincipalValid,
-  } = useContacts();
-  const { assets, getAssetIcon, asciiHex } = GeneralHook();
+  } = useCreateContact();
+
+  async function onAllowanceNewContactCheck() {
```
Comment (OliverSDSharp) [inline frontend/pages/contacts/components/addContact.tsx]: lets reduce this components size. too big with many funcionalities. lets move all functions to a hook.

File: frontend/pages/contacts/components/tableSubAccounts.tsx [hunk 4]
```diff
@@ -294,15 +300,17 @@ const TableSubAccounts = ({
     </table>
   );
 
-  function checkSubAcc(edit: boolean, cntc: Contact, asst: AssetContact, sa?: SubAccountContact) {
+  async function checkSubAcc(edit: boolean, cntc: Contact, asst: AssetContact, sa?: SubAccountContact) {
```
Comment (OliverSDSharp) [inline frontend/pages/contacts/components/tableSubAccounts.tsx]: lets also reduce the size of this component. moving logic to other file.

File: frontend/redux/CheckAuth.tsx [hunk 5]
```diff
@@ -91,15 +93,8 @@ export const handleLoginApp = async (authIdentity: Identity, fromSeed?: boolean)
...
-
-  // CONTACTS
-  const contactsData = localStorage.getItem("contacts-" + authIdentity.getPrincipal().toString());
-  if (contactsData) {
-    const contactsDataJson = JSON.parse(contactsData);
-    store.dispatch(setContacts(contactsDataJson.contacts));
-  }
-
-  // ALLOWANCES
+  console.log(principal);
```
Comment (OliverSDSharp) [inline frontend/redux/CheckAuth.tsx]: delete

**Review by OliverSDSharp — CHANGES_REQUESTED** [review]

File: frontend/pages/contacts/components/AddContact/SubAccountFormItem.tsx (hunk 2 above)
Comment (Julio-SDSharp) [reply frontend/pages/contacts/components/AddContact/SubAccountFormItem.tsx]: Done

File: frontend/helpers/icrc.ts (hunk 1 above)
Comment (Julio-SDSharp) [reply frontend/helpers/icrc.ts]: Done

File: frontend/pages/contacts/components/addContact.tsx (hunk 3 above)
Comment (Julio-SDSharp) [reply frontend/pages/contacts/components/addContact.tsx]: done

File: frontend/pages/contacts/components/tableSubAccounts.tsx (hunk 4 above)
Comment (Julio-SDSharp) [reply frontend/pages/contacts/components/tableSubAccounts.tsx]: done

File: frontend/redux/CheckAuth.tsx (hunk 5 above)
Comment (Julio-SDSharp) [reply frontend/redux/CheckAuth.tsx]: Done

File: frontend/components/modal/Modal.tsx [hunk 6]
```diff
@@ -34,15 +38,30 @@ const DialogDemo = ({
...
+              >
+                {isLoading && <LoadingLoader className="mr-2" />}
+                Yes
+              </button>
+            </Dialog.Close>
+          )}
+          {onOpenChange && (
             <button
-              className={`bg-RadioCheckColor rounded-lg py-2 ${disabled ? "opacity-50 pointer-events-none" : ""}`}
+              className={`bg-RadioCheckColor rounded-lg py-2 flex justify-center items-center ${
```
Comment (OliverSDSharp) [inline frontend/components/modal/Modal.tsx:56]: on this type of validations. use clxc. becuase this adds empty space to the class string

File: frontend/pages/contacts/components/AddContact/AddAssetOnCreate.tsx [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+export default function AddAssetOnCreate(props: AddAssetOnCreateProps) {
+  const { t } = useTranslation();
+  const { isValidSubacc, setNewContact, assets, newContact, selAstContact, newSubAccounts, setNewSubaccounts } = props;
+
+  return (
+    <div className="flex flex-col justify-start items-start w-[70%] h-full">
+      <div className="flex flex-row items-center justify-between w-full p-3">
+        <p className="whitespace-nowrap">{t("add.assets")}</p>
+        {assets.filter((ast) => {
```
Comment (OliverSDSharp) [inline frontend/pages/contacts/components/AddContact/AddAssetOnCreate.tsx:27]: maybe is best to place this o a usememo that way wont compute on each re-render.

File: frontend/pages/contacts/components/contactAssetPop.tsx [hunk 8]
```diff
@@ -45,12 +45,13 @@ const ContactAssetPop = ({
       <div className={`${compClass}`}>
         <DropdownMenu.Root open={openDrop && assets.length > 0} onOpenChange={setOpenDrop}>
           <DropdownMenu.Trigger className="!p-0">
-            <CustomButton
-              className={`${assets.length === 0 ? "bg-GrayColor cursor-default" : ""} ${btnClass}`}
-              size={"icon"}
+            <div
+              className={`h-[2rem] rounded-r-sm w-7 grid place-items-center ${
```
Comment (OliverSDSharp) [inline frontend/pages/contacts/components/contactAssetPop.tsx:49]: change for clxs

File: frontend/pages/contacts/components/tableContacts.tsx [hunk 9]
```diff
@@ -250,6 +259,7 @@ const TableContacts = ({
   function getContactsToShow() {
```
Comment (OliverSDSharp) [inline frontend/pages/contacts/components/tableContacts.tsx]: too much functions on component. maybe the best is to add them to the hook?

File: frontend/pages/home/components/ICRC/detail/transaction/SendForm/AllowanceContactBook.tsx [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        contact.contactName.toLocaleLowerCase().includes(cleanSearchValue) ||
+        contact.assetName?.toLocaleLowerCase().includes(cleanSearchValue) ||
+        contact.subAccountName?.toLocaleLowerCase().includes(cleanSearchValue)
+      );
+    });
+  }, [contacts, sender, searchSubAccountValue]);
+
+  return (
+    <div className="mx-4">
+      <DropdownMenu.Root modal={isOpen} onOpenChange={onOpenChange}>
```
Comment (OliverSDSharp) [inline frontend/pages/home/components/ICRC/detail/transaction/SendForm/AllowanceContactBook.tsx:70]: (i cannot add it to folder) lets split that many files, into diferent folders maybe. i see sender, receiver, contacts componets. all can be organized better.

File: frontend/pages/home/components/ICRC/detail/transaction/SendForm/ReceiverOwnerOld.tsx [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+        </DropdownMenu.Portal>
+      </DropdownMenu.Root>
+      <div className="flex items-center justify-center w-full"></div>
+      <p className="w-full text-left opacity-60">{t("to")}</p>
+
+      <div className={clsx(sendBox, "border-BorderSuccessColor")}>
+        <img src={SuccesIcon} alt="success-icon" />
+        <div className={clsx(accountInfo)}>
+          <p className="w-full pr-1 text-left break-words">
+            {contactToSend
```
Comment (OliverSDSharp) [inline frontend/pages/home/components/ICRC/detail/transaction/SendForm/ReceiverOwnerOld.tsx:149]: to complicated. lets avoid big ternary. can be on a function?

File: frontend/pages/home/components/ICRC/detail/transaction/SendForm/ReceiverOwnerOld.tsx [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+        <button
+          className="flex items-center justify-center p-1 rounded cursor-pointer bg-RadioCheckColor"
+          onClick={onMaxAmount}
+        >
+          <p className="text-sm text-PrimaryTextColor">{t("max")}</p>
+        </button>
+        <ExchangeIcon />
+      </div>
+      <div className="flex flex-row items-center justify-between w-full">
+        {!maxAmount().valid ? (
```
Comment (OliverSDSharp) [inline frontend/pages/home/components/ICRC/detail/transaction/SendForm/ReceiverOwnerOld.tsx:187]: too complex with ternaries. lets put it on a function

File: frontend/pages/home/components/ICRC/detail/transaction/SendForm/ReceiverType.tsx [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+import { clearReceiverAction, setReceiverIsManualAction } from "@redux/transaction/TransactionActions";
+import { useTranslation } from "react-i18next";
+
+export default function ReceiverType() {
+  const { t } = useTranslation();
+  const { receiver } = useAppSelector((state) => state.transaction);
+
+  return (
+    <div className="flex items-center justify-between w-full px-4 py-1">
+      <p className="font-bold opacity-50 text-PrimaryTextColorLight dark:text-PrimaryTextColor text-md">To</p>
```
Comment (OliverSDSharp) [inline frontend/pages/home/components/ICRC/detail/transaction/SendForm/ReceiverType.tsx:13]: translation

File: frontend/redux/Store.ts [hunk 14]
```diff
@@ -4,6 +4,7 @@ import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";
 import RootReducer from "./RootReducer";
 
 const store = configureStore({
+  devTools: true,
```
Comment (OliverSDSharp) [inline frontend/redux/Store.ts:7]: this does not go to dev/main

**Review by OliverSDSharp — CHANGES_REQUESTED** [review]

File: frontend/defaultTokens.ts [hunk 15]
```diff
@@ -13,6 +14,7 @@ export const ICRC1systemAssets: Array<Token> = [
     shortDecimal: "8",
     fee: "10000",
     subAccounts: [{ numb: "0", name: "Default", amount: "0", currency_amount: "0" }],
+    supportedStandards: [SupportedStandardEnum.Values["ICRC-1"], SupportedStandardEnum.Values["ICRC-2"]],
```
Comment (OliverSDSharp) [inline frontend/defaultTokens.ts]: If is all the same why not just 1 variable and put in on each supportedStandars?

**Review by OliverSDSharp — CHANGES_REQUESTED** [review]

**Review by OliverSDSharp — APPROVED** [review]

**Pull request merged**